[release/7.0-rc1] Add support for running final conventions during build

captainsafia opened the pull request:
## Description

This PR adds support for running conventions on an endpoint builder after "standard conventions" have been run. This allows these conventions to inspect metadata that has been registered by standard conventions and react to it. In particular, it allows groups to examine metadata of their endpoints.

Addresses https://github.com/dotnet/aspnetcore/issues/42750

## Customer Impact

The changes in this PR address a bug that made it impossible for users to leverage the route groups feature, `WithOpenApi`, and metadata annotations on their endpoints.

```csharp
var clients = app.MapGroup("/clients/{id}")
  .WithOpenApi();

clients.MapGet("/", () => ...)
  .Produces<Client>(); // Annotation doesn't appear in OpenApi definition without this change
```

## Regression?

- [ ] Yes
- [X] No

## Risk

- [ ] High
- [ ] Medium
- [X] Low

Low risk because:
- Changes is limited to ASP.NET Core surface area
- Change is only used in APIs introduced in .NET 7
- Change is additive and doesn't alter existing behavior

## Verification

- [X] Manual (required)
- [X] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [X] N/A

captainsafia committed:
Add support for running final conventions during build

halter73 commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -225,6 +229,19 @@ internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
+            {
+                groupFinalConventions[i](builder);
+            }
+        }
```
It would be convenient if we could run the final conventions after running the normal conventions for *all* the endpoints produced by this data source. That way conventions could be made aware of all the endpoints provided by the EndpointDataSource before finalizing the endpoints if the care. And I don't see a downside to running them this way.

captainsafia replied:
I think this is punt-worthy given the ✨ deadline ✨  but I've captured an issue for it in https://github.com/dotnet/aspnetcore/issues/43246.

halter73 replied:
Pro PR move 😆. I cannot come up with a real scenario for this yet, so this seems fair. It was more that if it were easy enough, it would be nice to have.

halter73 commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -265,6 +282,7 @@ private struct RouteEntry
         public IEnumerable<string>? HttpMethods { get; init; }
         public RouteAttributes RouteAttributes { get; init; }
         public ThrowOnAddAfterEndpointBuiltConventionCollection Conventions { get; init; }
+        public List<Action<EndpointBuilder>> FinalConventions { get; init; }
```
We should probably be as strict with throwing for normal conventions and final conventions.

halter73 commented on src/Http/Routing/src/DefaultEndpointConventionBuilder.cs:
```diff
@@ -11,11 +11,13 @@ internal sealed class DefaultEndpointConventionBuilder : IEndpointConventionBuil
     private List<Action<EndpointBuilder>>? _conventions;
+    private List<Action<EndpointBuilder>>? _finalConventions;
 
     public DefaultEndpointConventionBuilder(EndpointBuilder endpointBuilder)
```
Does anything still use this other than tests? I might have just forgotten to delete it when I moved all the old `Map` methods to `RouteHandlerBuilder`.

captainsafia replied:
Yep, this is only used in the tests.

halter73 commented on src/Mvc/Mvc.RazorPages/src/Infrastructure/PageActionEndpointDataSource.cs:
```diff
@@ -40,15 +40,15 @@ internal sealed class PageActionEndpointDataSource : ActionEndpointDataSourceBas
             if (actions[i] is PageActionDescriptor action)
             {
-                _endpointFactory.AddEndpoints(endpoints, routeNames, action, Array.Empty<ConventionalRouteEntry>(), conventions, CreateInertEndpoints);
+                _endpointFactory.AddEndpoints(endpoints, routeNames, action, Array.Empty<ConventionalRouteEntry>(), conventions, finalConventions, CreateInertEndpoints);
```
I guess I need to finish reviewing #42557 so this PR can react to the better group support.

captainsafia replied:
Yep, now that it's been merged I'll have to do quite a bit of rebasing on this PR. 😅

davidfowl commented:
This change looks nice

Kahbazi commented on src/Http/Http.Abstractions/src/Extensions/IEndpointConventionBuilder.cs:
```diff
@@ -16,4 +16,10 @@ public interface IEndpointConventionBuilder
+    /// Registers the specified convention for execution after the target builder has been built.
+    /// </summary>
+    /// <param name="finalConvention">The final convention to add to the builder.</param>
+    void Finally(Action<EndpointBuilder> finalConvention);
```
❓ 
```suggestion
    void Finally(Action<EndpointBuilder> finalConvention) { throw new NotSupportedException(); }
```

halter73 commented on src/Http/Http.Abstractions/src/Extensions/IEndpointConventionBuilder.cs:
```diff
@@ -16,4 +16,10 @@ public interface IEndpointConventionBuilder
+    /// Registers the specified convention for execution after the target builder has been built.
+    /// </summary>
+    /// <param name="finalConvention">The final convention to add to the builder.</param>
+    void Finally(Action<EndpointBuilder> finalConvention) => throw new NotSupportedException();
```
```suggestion
    void Finally(Action<EndpointBuilder> finalConvention) => throw new NotImplementedException();
```

BrennanConroy commented on src/Http/Routing/src/RouteGroupContext.cs:
```diff
@@ -11,39 +11,35 @@ namespace Microsoft.AspNetCore.Routing;
+    /// via <see cref="IEndpointConventionBuilder.Add(Action{EndpointBuilder})"/>. These should be applied in LIFO order when building every <see cref="RouteEndpoint"/>
+    /// returned from <see cref="EndpointDataSource.GetGroupedEndpoints(RouteGroupContext)"/>.
+    /// </summary>
+    public IReadOnlyList<Action<EndpointBuilder>> FinallyConvnentions { get; init; } = Array.Empty<Action<EndpointBuilder>>();
```
```suggestion
    public IReadOnlyList<Action<EndpointBuilder>> FinallyConventions { get; init; } = Array.Empty<Action<EndpointBuilder>>();
```

Needs to be fixed everywhere 😄

captainsafia replied:
*roblox death sound*

I'll take the L on this...

BrennanConroy commented on src/Http/Http.Abstractions/src/Extensions/IEndpointConventionBuilder.cs:
```diff
@@ -16,4 +16,10 @@ public interface IEndpointConventionBuilder
+    /// Registers the specified convention for execution after the target builder has been built.
+    /// </summary>
+    /// <param name="finalConvention">The final convention to add to the builder.</param>
+    void Finally(Action<EndpointBuilder> finalConvention) => throw new NotImplementedException();
```
Are we worried that making the default method throw will break existing scenarios? Basically, anywhere in the framework where we now call `Finally`, is the `IEndpointConventionBuilder` potentially a user defined one? And if so, was the code path accessible prior to 7.0?

captainsafia replied:
> Are we worried that making the default method throw will break existing scenarios?

This is a new API so I'm not too anxious about it breaking existing scenarios and the one place where it is used is `WithOpenApi` which was introduced in 7.0.

> Basically, anywhere in the framework where we now call Finally, is the IEndpointConventionBuilder potentially a user defined one?

Yep, I've added implementations in all the `IEndpointConvenntionBuilder` implementations except those that are used for tests:

- `ComponentEndpointConventionBuilder`
- `RouteHandlerBuilder`
- `ConnectionEndpointRouteBuilder`
- `RouteGroupBuilder`
- `HubEndpointConventionBuilder`
- `ControllerActionEndpointConventionBuilder`
- `PageActionEndpointConventionBuilder`

davidfowl replied:
Do a grep.app on github.

captainsafia replied:
> Do a grep.app on github.

YARP and gRPC also have implementations of IEndpointConventionBuilder, so we'll want to implement `Finally` there too.

However, at least for the time being, there's no risk of breaking existing scenarios in those APIs since they don't use it currently.

BrennanConroy replied:
It's not about gPRC or YARP or anyone elses implementation using `Finally`, it's about **our** code calling `Finally` on them.

captainsafia replied:
> It's not about gPRC or YARP or anyone elses implementation using Finally, it's about our code calling Finally on them.

Yeah, the point above about `WithOpenApi` stands. With this PR, it's the only place that might call into their `Finally`. It's not likely that users will be calling `WithOpenApi` on those builders though and, in the event that they are, it's probably a good thing that NIE is thrown since they don't fully support that API (intentionally).

halter73 replied:
Given that `WithOpenApi` already skips over non `RouteEndpointBuilder` endpoints and endpoints without `MethodInfo` metadata, skipping over unsupported endpoints is already a behavior. Should we swallow the `NotImplemented` exception from `Finally` in `WithOpenApi`? Also, I don't like `finalConvention` because it implies there can only be one of them.

```suggestion
    void Finally(Action<EndpointBuilder> finallyConvention) => throw new NotImplementedException();
```

captainsafia replied:
> Should we swallow the NotImplemented exception from Finally in WithOpenApi?

I'm not too enthused about treating the exception thrown here as some sort of contract that `WithOpenApi` will check for.

> endpoints without MethodInfo metadata,

It doesn't generate an operation for these types of endpoints but the `Finally` convention will still be registered on them so it's not totally the same.

> Also, I don't like finalConvention because it implies there can only be one of them.

Bleh. This wan't intentional. I used VS Code to do a find/replace of final => finally in the `src/Http` and `src/Mvc` directories but it looks like this one might've been skipped. :/

davidfowl replied:
Or maybe just call it convention because that also makes sense 😄. The property should be final conventions and the parameter should be convention. They are both adding conventions that run a different times.

davidfowl replied:
Imagine this was an abstract base class, the options are:
- Add a bool SupportsFinally, return false by default, make the new virtual member throw, or have a reasonable implementation (like calling Add?)

halter73 commented on src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs:
```diff
@@ -269,6 +266,21 @@ internal sealed class ActionEndpointFactory
+            finallyConventions[i](builder);
+        }
+
+        for (var i = groupFinallyConventions.Count - 1; i >= 0; i--)
```
Are these getting double reversed now? We need a test for this ordering if so.

halter73 commented on src/Http/Routing/src/EndpointDataSource.cs:
```diff
@@ -73,6 +73,11 @@ public virtual IReadOnlyList<Endpoint> GetGroupedEndpoints(RouteGroupContext con
 
+            for (var idx = context.FinallyConvnentions.Count - 1; idx >= 0; idx--)
+            {
+                context.FinallyConvnentions[idx](routeEndpointBuilder);
```
Are we also double reversing here?

halter73 commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -225,6 +229,22 @@ internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
+        {
+            // Group conventions are ordered by the RouteGroupBuilder before
+            // being provided here.
+            foreach (var groupFinallyConvention in groupFinallyConventions)
```
We got it right here at least. I'm now wondering about the wisdom of pre-reversing this list. It'd make more sense if the entry-specific finally conventions were stored in the stack instead of a List.

I no longer care where we reverse these. But we need regression tests that make sure we always run group and endpoint-specific `Finally` callbacks for all the `EndpointDataSource`'s we've modifed.

captainsafia replied:
> but we do want the LIFO ordering. But we need regression tests that make sure we always run group and endpoint-specific Finally callbacks in LIFO order for all the EndpointDataSource's we've modifed.

`FinallyONestednGroups_CanExamineFinallyInLIFOOrder` does this. I've updated it to include finally conventions on nested groups _and_ multiple finally conventions on an endpoint.

halter73 commented on src/Http/Routing/src/RouteGroupBuilder.cs:
```diff
@@ -78,12 +86,25 @@ public void Dispose()
             var combinedConventions = RoutePatternFactory.CombineLists(conventions, _routeGroupBuilder._conventions);
-            return new RouteGroupContext(fullPrefix, combinedConventions, applicationServices);
+            finallyConventions.Reverse();
+            _routeGroupBuilder._finallyConventions.Reverse();
```
Just calling `GetGroupedEndpoints()` shouldn't mutate the state. `GetGroupedEndpoints()` could be called multiple times. We should add a test where we add multiple levels of group finally conventions and call `GetGroupedEndpoints()` multiple times and make sure they all run in the right order.

captainsafia replied:
Good call. Added a test to validate idempotency of this operation.

captainsafia committed:
Fix typo and use stack for backing store

captainsafia commented:
585e4ecc47acbf3317f65825ff9a412f56decb40 contains the following:

- Fixes the typo in the renames for "finally conventions"
- Updates `RouteGroupBuilder`, `RouteEndpointDataSource`, and `ActionEndpointDataSource` implementations to use `Stack` to represent "finally conventions"
- Remove reversals and reverse-iterations in `Build` implementations

BrennanConroy commented on src/Http/Routing/test/UnitTests/Builder/RouteHandlerEndpointRouteBuilderExtensionsTest.cs:
```diff
@@ -1026,6 +1026,62 @@ public void RouteHandlerContext_ThrowsArgumentNullException_ForApplicationServic
+    }
+
+    [Fact]
+    public void FinallyONestednGroups_CanExamineFinallyInLIFOOrder()
```
```suggestion
    public void FinallyOnNestedGroups_CanExamineFinallyInLIFOOrder()
```

BrennanConroy commented on src/Http/Routing/test/UnitTests/Builder/RoutingEndpointConventionBuilderExtensionsTest.cs:
```diff
@@ -9,6 +9,26 @@ namespace Microsoft.AspNetCore.Builder;
 public class RoutingEndpointConventionBuilderExtensionsTest
 {
+    [Fact]
+    public void RequireHost_HostNames()
```
Seems random, what exactly is this test verifying? Maybe needs a more descriptive name

captainsafia replied:
This is a test that I moved over from [this file](https://github.com/dotnet/aspnetcore/blob/ac4bd5ef1660dbbd380de1e80c36468cc912770d/src/Http/Routing/test/UnitTests/RoutingEndpointConventionBuilderExtensionsTests.cs) that contained this singular test. It seemed weird to have this test on its own when it was testing the same surface area as these so I moved it over. I can take a stab at a better name now that it is in the larger file.

BrennanConroy commented on src/Mvc/Mvc.RazorPages/src/Infrastructure/DefaultPageLoader.cs:
```diff
@@ -88,6 +88,8 @@ private async Task<CompiledPageActionDescriptor> LoadAsyncCore(PageActionDescrip
                     },
             },
+            groupFinallyConventions: new Stack<Action<EndpointBuilder>>(),
+            finallyConventions: new Stack<Action<EndpointBuilder>>(),
```
The order of `groupFinally` vs `finally` keeps switching for different APIs, really triggering my OCD

BrennanConroy commented on src/Http/Routing/src/RouteGroupBuilder.cs:
```diff
@@ -109,5 +128,29 @@ private IChangeToken GetCompositeChangeToken()
         }
+
+        internal static IReadOnlyList<Action<EndpointBuilder>> CombineConventions(
+        IEnumerable<Action<EndpointBuilder>> firstColleciton,
+        IEnumerable<Action<EndpointBuilder>> secondCollection)
+        {
+            var firstSize = firstColleciton.Count();
+            var secondSize = secondCollection.Count();
+            var combinedConventions = new Action<EndpointBuilder>[firstSize + secondSize];
+
+            var i = 0;
+            foreach (var item in firstColleciton)
```
```suggestion
        IEnumerable<Action<EndpointBuilder>> firstCollection,
        IEnumerable<Action<EndpointBuilder>> secondCollection)
        {
            var firstSize = firstCollection.Count();
            var secondSize = secondCollection.Count();
            var combinedConventions = new Action<EndpointBuilder>[firstSize + secondSize];

            var i = 0;
            foreach (var item in firstCollection)
```

captainsafia commented on src/OpenApi/test/OpenApiRouteHandlerBuilderExtensionTests.cs:
```diff
@@ -116,7 +117,31 @@ public void WithOpenApi_WorksWithMapGroup()
+        string GetString() => "Foo";
+        var myGroup = builder.MapGroup("/group");
+        myGroup.WithOpenApi();
+        myGroup.MapDelete("/a", GetString).Produces<string>(201);
```
`WithOpenApi_GroupMetadataCanExamineAndExtendMoreLocalMetadata` covers this but I can add a more explicit test.

captainsafia committed:
Update application order and address feedback

- Fix ordering of group and entry-specific conventions in parameter list
- Update to use FIFO ordering within finally conventions of the same level
- Add tests for WithOpenApi on group and endpoint
- Revert to using List instead of Stack for convention
- Fix ordering of outer and inner group in RouteGroups

captainsafia commented:
3f1e1cddd35ab7ba8e37ae5db30612b89d6189c3 includes the following:

- Fix ordering of group and entry-specific conventions in parameter list
- Update to use FIFO ordering within finally conventions of the same level
- Add tests for WithOpenApi on group and endpoint
- Revert to using List instead of Stack for convention
- Fix ordering of outer and inner group in RouteGroups

halter73 reviewed (approved):
:shipit: ~(Maybe remove `required` from `RouteGroupContext.Prefix` though).~

Commit:
Add new test

captainsafia commented on src/Http/Routing/src/RouteGroupContext.cs:
```diff
@@ -11,39 +11,35 @@ namespace Microsoft.AspNetCore.Routing;
     /// not just the prefix supplied to the innermost call to <see cref="EndpointRouteBuilderExtensions.MapGroup(IEndpointRouteBuilder, RoutePattern)"/>.
     /// </summary>
-    public RoutePattern Prefix { get; }
+    public required RoutePattern Prefix { get; init; }
```
> An empty prefix seems to me every bit as reasonable as an empty service provider or empty conventions as a default value.

I was inclined to think this at first and also motivated to keep the property definitions consistent throughout this type.

Then, I figured that the prefix is an intrinsic and necessary part of the route group and it makes sense to force the API consumer to make an intentional decision about what it should be. In that way, I think it's different from the conventions and service providers.

Commit:
Update tests

captainsafia commented:
@dotnet/aspnet-admins Can I get help merging this?

Failing code check is because this PR introduces new APIs. 🙊
